tools/jgraph/jgraph.1
.\" SCCSID: @(#)jgraph.1	1.1	10/23/89
.\" SCCSID: @(#)jgraph.1	1.1	10/23/89
.TH jgraph 1
.SH NAME
jgraph \- filter for graph plotting to postscript
.SH SYNTAX
.B jgraph 
[\-\fIp\fR\|]
[\-\fIP\fR\|]
[\-\fIL\fR\|]
[\-\fIcomments\fR\|]
[\fIfilename\fR ...\|]
.SH DESCRIPTION
\fBJgraph\fR
takes the description of a graph or graphs
and produces a postscript file on the standard output.  
\fBJgraph\fR
is ideal for plotting any mixture of scatter point graphs, line
graphs, and/or bar graphs, and embedding the output into LaTeX, or
any other text processing system which can read postscript.
.sp
\fBJgraph\fR reads its input from the specified files.  If no
files are specified, then it reads from standard input.
.sp
The graph description language is simple enough to get nice looking
graphs with a minimum of effort, yet powerful enough to give the user
the flexibility to tailor the appearance of the graph to his or her
individual preferences.  This includes plotting multiple graphs and
laying them out separately on the page (or pages).
.sp
As an example, if the user wanted to simply plot the points (2,3),
(4,5), (1,6), the following would be enough of a specification file:
.PP
.nf
	newgraph
	newcurve pts 2 3 4 5 1 6
.fi
.PP
Now, if the user wanted to spruce the graph up by adding labels to
the axes, connecting the points, and titling the graph, then the
input could change to:
.PP
.nf
	newgraph
	newcurve pts 2 3 4 5 1 6 linetype solid
	xaxis label : X axis
	yaxis label : Y axis
	title : This is an example graph
.fi
.PP
If the user instead wanted this to be a bar graph with different
endpoints on the axes, he/she could simply change the input to:
.PP
.nf
	newgraph
	xaxis min 0 max 5 label : X axis
	yaxis min 0 max 6 label : Y axis
	newcurve pts 2 3 4 5 1 6 marktype xbar
	title : This is an example bar graph
.fi
.PP
There are many more features of the description language, which are
described below in the next section.  Features which are not embedded
within the description language are: line and function interpolation,
function plotting, and pie graphs.  The latter is impossible to do
with the aid of 
\fBjgraph, \fR
however, the others can be effected with
\fBjgraph \fR
mixed with awk or c.  See 
FUNCTION PLOTTING AND OTHER NON-INHERENT FEATURES 
below.
.sp
Also below is a section 
HINTS AND EXAMPLE GRAPHS, which may give good
ideas on how to use 
\fBjgraph \fR
more effectively.
.SH OPTIONS
.TP
.B \-P
The 
\fB\-P\fR
option produces postscript which can be piped directly to 
\fBlpr,\fR  
which can be displayed in an Xwindows environment with
\fBgs\fR
(ghostscript).
Without this option, the output should be embedded within 
\fBLaTeX\fR
or a similar text processing system.
.TP
.B \-L
The 
\fB\-L\fR
option produces a landscape plot.
.TP
.B \-p
The 
\fB\-p\fR
option re-prints the input on the standard output, only
with all the defaults made explicit.  This is useful for letting the
user do his/her own special formatting, as it shows the explicit
values that the defaults assume, so that they can be manipulated.
.TP
.B \-comments
This option makes jgraph put comments into the output postscript.  These
make it easier for the user to wade through the final postscript if 
necessary. 
.SH THE DESCRIPTION LANGUAGE
The description language is essentially keywords followed by
attributes.  All keywords and attributes except for string attributes
are tokens -- non-white-space characters surrounded by white-space. 
Special tokens are 
``(*'', ``*)'', ``include'', ``:'', and ``shell'', which denote
comments, include-file statements, string identifiers, and shell-include
statements:
.TP 
.B Comments
Comments are surrounded by the tokens ``(*'' ``*)'' as in
Modula-2 (except that here, the tokens must be surrounded by white-
space).  Comments may be nested.  If the comment runs to the end of a
file, the last ``*)'' may be omitted.
.TP 
.B Include\-file statements
The token following an ``include'' token is
expected to be a file name.  The result of the statement is to
include the contents of the file at that point.  Include-file
statments can be nested within included files, and within shell
includes.
.TP 
.B Strings
In places where strings are required (as in graph and
curve labels), they are denoted by the token ``:''.  The second
character after the ``:'' starts the string, and the next newline
character terminates it.  
Thus, the string ``Graph #1'' can be denoted as:
.nf

	: Graph #1<newline>

or

	:<newline>
	Graph #1<newline>

.fi
One can get multiline strings by making
a backslash the last character before the newline on all but the
last line.  Notice that in strings white-space is not ignored.  
This way of denoting strings allows the user to embed leading and
trailing spaces, as well as the null string.  For example, the
null string ``'' is represented by:
.nf

	: <newline>

.fi
Once a string has been started, it may contain any character.
Specifically, it may contain the sequence ``(*'', ``shell'', 
or ``include'' without starting a comment or including a file.
Each line of a string must contain less than 1000 characters.  Otherwise
string sizes are limited only by the size of memory.
.TP
.B Shell\-include statements
Shell include statements are of the form ``shell'', ``:'', and then
a string.  The result of the statement is that the string is executed
(using popen, which passes the string to sh), and the standard
output is included at that point.  Shell-includes can be freely
nested within include-files and other shell-includes.  Shell 
commands may be more than one line, but must not exceed 1000 characters.
The shell statement is not (yet) available on VMS.
.TP
.B Notation
In the descriptions below: 
.RS
.TP
\fBtk \|{\fIinteger\fB\|}\fR
means that token
\fBtk \fR
must be followed by an integer.  
.TP
\fBtk \|[\fIinteger\fB\|]\fR
means that 
\fBtk\fR
may be followed by an integer, but doesn't have to.  In most cases, if
\fBtk\fR
is not followed by an integer, then the command denoted by
\fBtk \fR
is ignored.  
.TP
\fBtk \|[\|{\fIinteger\fB\|} \|{\fIinteger\fB\|}\|]*
means that 
\fBtk\fR
must be
followed by an even number of integers.  
.PD
.LP
Supported types other than
integer are: 
\fB\|{\fIfloat\fB\|} \fR
for floating point entries, 
\fB\|{\fItoken\fB\|} \fR
for any
token, and 
\fB\|{\fIstring\fB\|} \fR
for a string as defined above.
.RE
.TP
.B TOP-LEVEL DESCRIPTION COMMANDS
.RS
.TP
.B newgraph
This starts editing a new graph (see GRAPH EDITING
COMMANDS).  Note that multiple graphs may be drawn on the same page.
.TP
\fBgraph \|{\fIinteger\fB\|}\fR
This edits the graph denoted by 
\fB\|{\fIinteger\fB\|}.  \fR
If the graph doesn't exist, then this command creates it and starts
editing it.  
\fBNewgraph\fR
is simply an abbreviation for 
\fB\fIgraph\fB \fIn\fB\fR
where n=0 if this is the first graph, otherwise n=m+1, where m is the
largest number of any graph so far.
.TP
\fBcopygraph \|[\fIinteger\fB\|]\fR
This creates a new graph, and copies all the attributes from the
graph
\fB\|[\fIinteger\fB\|]'s\fR
x and y axes, as well as its 
\fB\fIx_translate\fB\fR
and 
\fB\fIy_translate\fB\fR
values, the clipping, the legend defaults, and 
the title defaults.   If the
\fB\|[\fIinteger\fB\|]\fR
is omitted, then it copies its values from the ``previous''
graph, which is 
defined to be the graph with the largest number 
less than the currrent graph's number.  If the current 
graph has the smallest number, then it will take the last graph from 
the previous page of graphs.  If there is no previous page, then an 
error will be flagged.
(copygraph does not copy the values of the 
\fB\fIhash_at\fB, \fImhash_at\fB,\fR
and 
\fB\fI\fIhash_label\fB\fR 
attributes). 
.TP
.B newpage
This command is for plotting graphs on multiple pages.  After a 
\fBnewpage,\fR
the graphs that the user enters will be plotted on a new page.
New graphs and strings will be numbered starting with 0.  
Essentially, 
\fB\fInewpage\fB\fR
is the same as appending together the output of separate calls of
jgraph on the text before the 
\fB\fInewpage,\fB\fR
and on the text after the 
\fB\fInewpage.\fB\fR
\fB\fINewpage\fB\fR
will most likely produce bizarre results if the
\fB\-P\fR
option is not specified.
.TP
\fBX \|[\fIfloat\fB\|]\fR
.br
.ns
.TP
\fBY \|[\fIfloat\fB\|]\fR
Postscript files to be embedded in LaTeX (and some other programs)
contain a ``bounding box''
which defines the area which LaTeX will allocate for the postscript.
Other programs use this bounding box as well, sometimes using it
to define where to clip the postscript image.
\fBJgraph \fR
uses the axis lines and labels, and the title to generate its 
bounding box.  Most of the time that's good enough to work in 
LaTeX.  The 
\fB\fIY\fB\fR
and
\fB\fIX\fB\fR
commands say to make the height and width of the bounding box at least
\fB\fIY\fB\fR
and
\fB\fIX\fB\fR
inches, respectively, but to maintain the current centering of the 
graph.  If you still need further control over the
bounding box (e.g. to change the centering), try the
\fB\fIbbox\fB\fR
command.  If there's more than one page in the jgraph file, 
\fB\fIY,\fB\fR
\fB\fIX\fB\fR
and
\fB\fIbbox\fB\fR
values can be given for each graph.
.TP
\fBbbox \fIfloat\fB \fIfloat\fB \fIfloat\fB \fIfloat\fB\fR
If the 
\fB\fIY\fB\fR
and
\fB\fIX\fB\fR
commands aren't enough to help you define a good bounding box, this
command lets you explicitly enter one which will go directly into the 
jgraph output.  Its units are the 
final postscript units.  It's probably best to use the 
\fB\-p\FR 
option
to see what the bounding box is that jgraph produces, and then
alter that accordingly with
\fB\fIbbox.\fB\fR
The main use for this is to change the automatic centering that jgraph
performs:  Usually the center of the bounding box that jgraph computes 
is put at the center of the page.  Changing the bbox changes this 
center.
.TP
\fBpreamble : \|{\fIstring\fB\|}\fR   
.br
.ns
.TP
\fBpreamble {\fItoken\fB\|}\fR   
.br
.ns
.TP
\fBepilogue : \|{\fIstring\fB\|}\fR   
.br
.ns
.TP
\fBepilogue {\fItoken\fB\|}\fR   
These two commands allow the user to include strings or
files (the token specifies the filename) which will be copied directly 
into jgraph's output.
The \fIpreamble\fB is included at the beginning of the output
(after some initial postscript to set things up for jgraph), 
and the \fIepilogue\fB is included at the end.  A good use for
the \fIpreamble\fB is to set up a postscript dictionary if you're
using postscript marks.
.PD
.RE
.LP
.TP
.B GRAPH EDITING COMMANDS
These commands act on the current graph. 
Graph editing is terminated when one of the top-level description
commands is given.
.RS
.TP
\fBxaxis\fR
.br
.ns
.TP
\fByaxis\fR
Edit the x or y axis (see AXIS EDITING COMMANDS)
.TP
\fBnewcurve\fR  
This starts editing a new curve of the graph (see CURVE
EDITING COMMANDS).
.TP
\fBcurve \|{\fIinteger\fB\|}\fR   
This edits the curve denoted by 
\fB\|{\fIinteger\fB\|}.  \fR
If the curve doesn't exist, then this command creates it and starts
editing it.  
\fINewcurve\fB
and 
\fIcurve\fB
interact as 
\fInewgraph\fB
and
\fIgraph\fB
do.
.TP
\fBnewline\fR  
This is an abbreviation for:
.PP
.nf
	  newcurve marktype none linetype solid
.fi
.PP
.TP
\fBcopycurve \|[\fIinteger\fB\|]\fR
This starts editing a new curve of the graph, and copies all its 
values except for the points from curve 
\fB\|[\fIinteger.\fB\|]\fR
If the 
\fB\|[\fIinteger\fB\|]\fR
is omitted, then it copies its values from the 
last curve in this graph.  If this graph currently has
no curves, then it searches backwards from the previous graph.
.TP
\fBtitle\fR   
This edits the title of the graph (see LABEL EDITING
COMMANDS).  The title is given a default location centered beneath
the graph, and a default font size of 12, however, as with all
labels, this can be changed.
.TP
\fBlegend\fR   
The edits the legend of the graph (see LEGEND EDITING
COMMANDS).  As a default, the graph will contain a legend
if any of its curves have labels.
.TP
\fBnewstring\fR   
This edits a new text string (see LABEL EDITING
COMMANDS).  This is useful as it allows the user to plot text on the
graph as well as curves.
.TP
\fBstring \|{\fIinteger\fB\|}\fR   
.br
.ns
.TP
\fBcopystring \|[\fIinteger\fB\|]\fR
\fIString\fB
and
\fIcopystring\fB
are to 
\fInewstring\fB
as
\fIcurve\fB
and
\fIcopycurve\fB
are to
\fInewcurve.\fB
.TP
\fBborder\fR
.br
.ns
.TP
\fBnoborder\fR
\fIBorder\fR\fB
draws a square border around the area defined by the axes.
\fINoborder\fB
specifies no border.
\fINoborder\fB
is the default.  
.TP
\fBclip\fR
.br
.ns
.TP
\fBnoclip\fR
\fIClip\fB
specifies that all curves in the graph will be clipped -- that is,
no points outside of the of axes will be plotted.  Clipping can also be
specified on a per-curve basis.  The default is
\fInoclip.\fB
.TP
\fBinherit_axes\fR   
This is an old command which is kept for backward compatibility.
\fICopycurve.\fB
is equivalent to:
.PP
.nf
	  newgraph inherit_axes
.fi
.PP
.TP
\fBx_translate \|[\fIfloat\fB\|]\fR   
.br
.ns
.TP
\fBy_translate \|[\fIfloat\fB\|]\fR   
By default, the bottom left-hand corner of each graph is at point
(0,0) (final postscript units).
\fIX_translate\fB
and
\fIY_translate\fB
translate the bottom left-hand corner of the graph 
\fB\|[\fIfloat\fB\|] \fR
inches.  The main use of this is to draw more than one graph on 
a page.  Note that jgraph considers all the graphs drawn on the
page when it computes its bounding box for centering.  Thus, if
only one graph is drawn, it will always be centered on the page, 
regardless of its 
\fIX_translate\fB
and
\fIY_translate\fB
values.  These values are used for relative placement of the graphs.
   To change the centering of the graphs, use
\fIbbox.\fB
.TP
\fBX \|[\fIfloat\fB\|]\fR
.br
.ns
.TP
\fBY \|[\fIfloat\fB\|]\fR   
These are the same as 
\fIX\fB
and 
\fIY\fB
in the
Top-level commands, except that they let the user continue editing
the current graph.
.PD
.RE
.LP
.TP
.B SIMPLE AXIS EDITING COMMANDS
These commands act on the current
axis as chosen by 
\fIxaxis\fB
or
\fIyaxis\fB
(see GRAPH EDITING COMMANDS). 
Axis editing terminates when a graph or top-level command is given. 
There are more advanced axis editing commands given below which have
to do with moving the hash marks, adding new hash marks and labels,
etc.  See ADVANCED AXIS EDITING COMMANDS.
.RS
.TP
\fBlinear\fR
.br
.ns
.TP
.B log
Set the axis to be linear or logarithmic. The
default is linear.  If the axis is set to be logarithmic, then values
<= 0.0 will be disallowed, as they are at negative infinity on the
axis.
.TP
\fBmin \|[\fIfloat\fB\|]\fR
.br
.ns
.TP
\fBmax \|[\fIfloat\fB\|]\fR   
Set the minimum and maximum values of
this axis.  Defaults depend on the points given.  They can be seen by
using the 
\fB\-p \fR
option.  Unless stated, all units (for example point
plotting, string plotting, etc.) will be in terms of the
\fImin\fB
and
\fImax\fB
values of the x and y axes.
.TP
\fBsize \|[\fIfloat\fB\|]\fR   
Set the size of this axis in inches.
.TP
\fBlog_base \|[\fIfloat\fB\|]\fR   
Set the base of the logarithmic axis. Default =
10.  This is the value which determines which hash
marks and hash labels are automatically produced.
.TP
\fBhash \|[\fIfloat\fB\|]\fR   
Hash marks will be 
\fB\|[\fIfloat\fB\|] \fR
units apart.  Default = -1.  
If this value equals 0, then there will be no hash marks.  If
this value is less than 0, then the hash marks will be automatically
set by 
\fBjgraph \fR
(see 
\fB\-p \fR
for the value).  By default, each hash mark
will be labeled with its value.  
\fIHash\fB
and 
\fIshash\fB
are ignored if
the axes are logarithmic.
.TP
\fBshash \|[\fIfloat\fB\|]\fR   
Make sure there is a hash mark at the point
\fB\|[\fIfloat\fB\|] \fR
along the axis.  The default is set by 
\fBjgraph\fR
if
\fBhash\fR
= -1.
If 
\fIhash\fB
is set by the user, 
\fIshash\fB
is defaulted to the
\fImin\fB
value of the axis.
.TP
\fBmhash \|[\fIinteger\fB\|]\fR   
Put 
\fB\|[\fIinteger\fB\|] \fR
minor hash marks between the above
hash marks.  Default = -1.  If this value equals 0, then there will
be no minor hash marks.  If this value is negative, then the value
will be chosen by 
\fBjgraph \fR
(see 
\fB\-p\fR
for the value).
.TP
\fBprecision \|[\fIinteger\fB\|]\fR   
.TP
\fBhash_format \fItoken\fB\fR   
These control how jgraph formats the automatic hash labels.
The user shouldn't have to worry about these values, except in
extreme cases.  Jgraph uses \fBprintf\fR to format the labels.
If \fBhash_format\fR is ``f'' (the default), then the 
value of a hash label is printed with 
.PP
.nf
	printf("%.*f", precision, value).
.fi
.PP
Other valid \fBhash_format\fR values are ``G'', ``g'', ``E'', and ``e''.
``G'' is a good generic format which converts to scientific notation 
if the value becomes too big or too small.
If the precision is negative, then jgraph chooses a default:  For 
``g'' and ``G'', the default is 6.  For ``e'' and ``E'', the default
is 0, and for ``f'', jgraph tries to determine a reasonable default.
Please read the man page of \fBprinf(1)\fR for a complete description
of how it formats floating point numbers.
.TP
\fBlabel\fR   
Edit the label of this axis (see LABEL EDITING COMMANDS). 
By default, the label is in font ``Times-Bold'', and has a font size of
10.  If the user doesn't change any of the plotting attributes of the
label, 
\fBjgraph \fR
chooses an appropriate place for the axis label.
.TP
\fBdraw_at \|[\fIfloat\fB\|]\fR   
Draw the axis line at this point on the other axis. 
The default is usually the other axis's 
\fImin, \fB
however if 
\fIhash_scale \fB
is positive (see 
\fIhash_scale \fB
under ADVANCED AXIS EDITING), it will be
the other axis's 
\fImax.\fB
.TP
\fBnodraw\fR   
Do not draw the axis, the hash marks or any labels.  This
is useful for plotting points with no axes, and for overlaying graphs
on top of one another with no clashes.  This is equivalent to
\fIno_draw_axis,\fB
\fIno_draw_axis_label,\fB
\fIno_draw_hash_marks,\fB
and
\fIno_draw_hash_labels.\fB
.TP
\fBdraw\fR   
Cancels the effect of 
\fInodraw.  \fB
Default = 
\fIdraw.\fB
This is
equivalent to 
\fIdraw_axis,\fB
\fIdraw_axis_label,\fB
\fIdraw_hash_marks,\fB
and 
\fIdraw_hash_labels.\fB
.TP
\fBgrid_lines\fR
.br
.ns
.TP
\fBno_grid_lines\fR
\fIGrid_lines\fB
specifies to plot a grid line at each major hash
mark on this axis.  The default is 
\fIno_grid_lines.\fB
.TP
\fBmgrid_lines\fR
.br
.ns
.TP
\fBno_mgrid_lines\fR
\fIMgrid_lines\fB
specifies to plot a grid line at each minor hash
mark on this axis.  The default is 
\fIno_mgrid_lines.\fB
.PD
.RE
.LP
.TP
.B CURVE EDITING COMMANDS 
These commands act on the current curve as
chosen by 
\fInewcurve\fB
or 
\fIcurve\fB
(see GRAPH EDITING COMMANDS).  Curve
editing terminates when a graph or top-level command is given.
.RS
.TP
\fBpts \|[\|{\fIfloat\fB\|} \|{\fIfloat\fB\|}\|]*\fR   
This sets the points to plot in this
curve.  The first 
\fIfloat\fB
is the x value, and the second 
\fIfloat\fB
is the y
value of the point.  Points are plotted in the order specified.
This command stops reading points when a non-float is given.
The user can specify this command multiple times within a curve -- 
each time, simply more points are added to the curve.
.TP
\fBx_epts \|[\|{\fIfloat\fB\|} \|{\fIfloat\fB\|} \|{\fIfloat\fB\|} \|{\fIfloat\fB\|}\|]*\fR   
.br
.ns
.TP
\fBy_epts \|[\|{\fIfloat\fB\|} \|{\fIfloat\fB\|} \|{\fIfloat\fB\|} \|{\fIfloat\fB\|}\|]*\fR   
This allows the user to specify points and ``confidence values'' (otherwise
known as ``error bars'').  The first two 
\fIfloats\fB
specify the x and y values of
the point, as above.  If
\fBx_epts\fR
is specified,  then the second two
\fIfloats\fB
specify range or confidence values
for the x value of the point.  
Error bars will be printed to each of these x values (using the 
original point's y value)
from the original point.  Similarly,
\fIy_epts\fB
specifies range or confidence values for the y value of the point.
\fIpts\fB
\fIx_epts\fB
and
\fIy_epts\fB
can all be intermixed. 
.TP
\fBmarktype\fR   
This sets the kind of mark that is plotted for this curve.  Valid
marks are: \fIcircle\fR, \fIbox\fR, \fIdiamond\fR, \fItriangle\fR,
\fIx\fR, \fIcross\fR, \fIellipse\fR, \fIxbar\fR, \fIybar\fR,
\fItext\fR, \fIpostscript\fR, \fIeps\fR, \fInone\fR, and variants of
\fIgeneral\fR.  Most of these are self-explanatory, except for the
last few: 
  \fIXbar\fR makes the curve into a bar graph with the bars going
to the x axis.  \fIYbar\fR has the bars going to the y axis. 
  \fIText\fR lets the user plot text instead of a mark.  The text is 

tools/jgraph/jgraph.c
/* jgraph.c
 * James S. Plank
 
Jgraph - A program for plotting graphs in postscript.

 * $Source: /Users/plank/src/jgraph/RCS/jgraph.c,v $
 * $Revision: 8.5 $
 * $Date: 2017/11/28 17:33:27 $
 * $Author: plank $

James S. Plank
Department of Electrical Engineering and Computer Science
University of Tennessee
Knoxville, TN 37996
plank@cs.utk.edu

Copyright (c) 2011, James S. Plank
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

 - Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.

 - Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in
   the documentation and/or other materials provided with the
   distribution.

 - Neither the name of the University of Tennessee nor the names of its
   contributors may be used to endorse or promote products derived
   from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.
*/

#include <stdio.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


#ifdef VMS 
#include <redexp.VMS>
#endif

#include "jgraph.h"

int NMARKTYPES = 17;
int NORMALMARKTYPES = 6;

char *MARKTYPESTRS[] = { "circle", "box", "diamond", "triangle", "x", "cross", 
                         "ellipse", "general", "general_nf", "general_bez",
                         "general_bez_nf", "postscript", "eps", 
                         "xbar", "ybar", "none", "text"};
char MARKTYPES[] = {     'o',      'b',   'd',       't',        'x', 'c', 
			 'e',       'g',       'G',          'Z',
                         'B',              'p',          'E',
                         'X',    'Y',    'n',    'l' };

int NPATTERNS = 3;
char *PATTERNS[] = { "solid", "stripe", "estripe" };
char PTYPES[] = { 's', '/', 'e' };

Label new_label()
{
  Label l;
  int i;

  l = (Label) malloc (sizeof(struct label));
  l->label = CNULL;
  l->hj = 'c';
  l->vj = 'b';
  l->font = "Times-Roman";
  l->fontsize = 9.0;
  l->rotate = 0;
  l->graytype = 'n';
  for (i = 0; i < 3; i++) l->gray[i] = 0.0;
  l->linesep = FSIG;

  return l;
}


Curve new_curve(c, num)
Curve c;
int num;
{
  Curve new_c;
  int i;

  new_c = (Curve) get_node((List) c);
  new_c->num = num;
  new_c->l = new_label();
  new_c->l->vj = 't';
  new_c->lmark = new_label();
  new_c->lmark->hj = 'c';
  new_c->lmark->vj = 'c';
  new_c->clip = 0;
  new_c->pattern = 's';
  new_c->parg = FSIG;
  new_c->apattern = 's';
  new_c->aparg = FSIG;
  new_c->ppattern = 's';
  new_c->pparg = FSIG;
  new_c->graytype = 'n';
  for (i = 0; i < 3; i++) new_c->gray[i] = 0.0;
  new_c->afilltype = 'g';
  for (i = 0; i < 3; i++) new_c->afill[i] = 0.0;
  new_c->pts = (Point) make_list(sizeof(struct point));
  new_c->yepts = (Point) make_list(sizeof(struct point));
  new_c->xepts = (Point) make_list(sizeof(struct point));
  new_c->npts = 0;
  new_c->gen_linetype = (Flist) make_list(sizeof(struct flist));
  new_c->marktype = MARKTYPES[num % NORMALMARKTYPES];
  new_c->linetype = '0';
  new_c->linethick = 1.0;
  new_c->marksize[0] = FSIG;
  new_c->marksize[1] = FSIG;
  new_c->mrotate = 0.0;
  new_c->general_marks = (Point) make_list(sizeof(struct point));
  new_c->filltype = 'n';
  for (i = 0; i < 3; i++) new_c->fill[i] = 0.0;
  new_c->pfilltype = 'n';
  for (i = 0; i < 3; i++) new_c->pfill[i] = 0.0;
  new_c->poly = 0;
  new_c->rarrow = 0;
  new_c->larrow = 0;
  new_c->rarrows = 0;
  new_c->larrows = 0;
  new_c->asize[0] = FSIG;
  new_c->asize[1] = FSIG;
  new_c->bezier = 0;
  new_c->postscript = CNULL;
  new_c->postfile = 0;
  new_c->eps = CNULL;
  prio_insert(new_c, c, 0);
  return new_c;
}

Curve new_line(c, num)
Curve c;
int num;
{
  Curve new_c;
  new_c = new_curve(c, num);
  new_c->linetype = 's';
  new_c->marktype = 'n';
  return new_c;
}


Curve get_curve(c, num)
Curve c;
int num;
{
  Curve new_c;
  for(new_c = last(c); new_c != nil(c) && new_c->num > num; new_c = prev(new_c));
  if (new_c == nil(c) || new_c->num < num) return new_curve(c, num);
  return new_c;
}

Axis new_axis(is_x)
int is_x;
{
  int i;
  Axis a;

  a = (Axis) malloc (sizeof(struct axis));
  a->label = new_label();
  a->label->x = FSIG;
  a->label->y = FSIG;
  a->label->font = "Times-Bold";
  a->label->fontsize = 10.0;
  a->label->rotate = FSIG;
  a->label->hj = '0';
  a->label->vj = '0';
  a->size = 3.0;
  a->max = FSIG;
  a->min = FSIG;
  a->pmax = FSIG;
  a->pmin = FSIG;
  a->hash_interval = -1.0;
  a->log_base = 10.0;
  a->minor_hashes = -1;
  a->precision = -1;
  a->hl = new_label();
  a->hl->label = "";
  a->hl->font = "Times-Roman";
  a->hl->fontsize = 9.0;
  a->hl->rotate = 0.0;
  a->hl->hj = '0';
  a->hl->vj = '0';
  a->hash_format = 'f';
  a->draw_at = FSIG;
  a->draw_hash_marks_at = FSIG;
  a->draw_hash_labels_at = FSIG;
  a->draw_hash_labels = 1;
  a->draw_axis_line = 1;
  a->draw_hash_marks = 1;
  a->draw_axis_label = 1;
  a->auto_hash_labels = 1;
  a->auto_hash_marks = 1;
  a->start_given = 0;
  a->hash_scale = -1.0;
  a->grid_lines = 0;
  a->mgrid_lines = 0;
  a->graytype = 'n';
  for (i = 0; i < 3; i++) a->gray[i] = 0.0;
  a->gr_graytype = '0';
  for (i = 0; i < 3; i++) a->gr_gray[i] = 0.0;
  a->mgr_graytype = '0';
  for (i = 0; i < 3; i++) a->mgr_gray[i] = 0.0;
  a->is_x = is_x;
  a->is_lg = 0;
  a->hash_labels = (String) make_list (sizeof(struct string));
  a->hash_lines = (Hash) make_list (sizeof(struct hash));
  return a;
}

Legend new_legend()
{
  Legend l;
  l = (Legend) malloc (sizeof(struct legend));
  l->linelength = FSIG;
  l->linebreak = FSIG;
  l->midspace = FSIG;
  l->type = 'u';
  l->l = new_label();
  l->l->label = "";
  l->l->hj = 'l';
  l->l->vj = 'c';
  l->l->x = FSIG;
  l->l->y = FSIG;
  return l;
}

Label new_title()
{
  Label t;

  t = new_label();
  t->x = FSIG;
  t->y = FSIG;
  t->rotate = 0.0;
  t->hj = 'c';
  t->vj = 't';
  t->fontsize = 12.0;
  return t;
}

String new_string(s, num)
String s;
int num;
{
  String new_s;

  new_s = (String) get_node((List) s);
  new_s->num = num;
  new_s->s = new_label();
  prio_insert(new_s, s, 0);
  return new_s;
}

String get_string(s, num)
String s;
int num;
{
  String new_s;
  for(new_s = last(s); new_s != nil(s) && new_s->num > num; new_s = prev(new_s));
  if (new_s == nil(s) || new_s->num < num) return new_string(s, num);
  return new_s;
}

/*
Default new_default()
{
  Default new_d;
  new_d = (Default) malloc (sizeof(default));
  new_d->rotate = FSIG;
  new_d->fontsize = FSIG;
  new_d->general_marks = GMNULL;
  new_d->fill = FSIG;
  new_d->linethick = FSIG;
  new_d->font = CNULL;
  new_d->hj = '!';
  new_d->vj = '!';
  new_d->marktype = '!';
  new_d->linetype = '!';
  new_d->marksize[0] = FSIG;
  new_d->marksize[1] = FSIG;
}
*/
  
Graph new_graph(gs, num)
Graph gs;
int num;
{
  Graph g;

  g = (Graph) get_node((List) gs);
  g->num = num;
  g->xminval = 0.0;
  g->yminval = 0.0;
  g->xmaxval = 0.0;
  g->ymaxval = 0.0;
  g->x_axis = new_axis(1);
  g->y_axis = new_axis(0);
  g->x_translate = 0.0;
  g->y_translate = 0.0;
  g->curves = (Curve) make_list(sizeof(struct curve));
  g->strings = (String) make_list(sizeof(struct string));
  g->title = new_title();
  g->clip = 0;
  g->border = 0;
  g->legend = new_legend();
/*  g->def = new_default(); */
  prio_insert(g, gs, 0);
  return g;
}

Graph get_graph(g, num)
Graph g;
int num;
{
  Graph new_g;
  for(new_g = last(g); new_g != nil(g) && new_g->num > num; new_g = prev(new_g));
  if (new_g == nil(g) || new_g->num < num) return new_graph(g, num);
  return new_g;
}

void new_graphs(gs)
Graphs gs;
{
  Graphs newg;

  newg = (Graphs) get_node((List) gs);
  newg->g = (Graph) make_list(sizeof(struct graph));
  newg->height = 0.0;
  newg->width = 0.0;
  newg->bb[0] = ISIG; newg->bb[1] = ISIG;
  newg->bb[2] = ISIG; newg->bb[3] = ISIG;
  newg->preamble = CNULL;
  newg->epilogue = CNULL;
  newg->prefile = 0;
  newg->epifile = 0;
  if (first(gs) == nil(gs)) newg->page = 1; else newg->page = last(gs)->page+1;
  insert((List) newg, (List) gs);
}

int main(argc, argv)
int argc;
char **argv;
{
  Graphs gs;
  int i;
  int show, pp;
  int landscape;
  int comments;
  int nfiles;

#ifdef VMS 
  IOSTREAM = stdin;	/* for token.c -hdd */
#endif
  show = 0;
  pp = 0;
  comments = 0;
  landscape = 0;
  nfiles = 0;

  for (i = 1; i < argc; i++) {
    if (strcmp(argv[i], "-p") == 0) show = 1;
    else if (strcmp(argv[i], "-comments") == 0) comments = 1;
    else if (strcmp(argv[i], "-P") == 0) pp = 1;
    else if (strcmp(argv[i], "-L") == 0) landscape = 1;
    else {
      nfiles++;
      set_input_file(argv[i]);
    }
  }
  if (nfiles == 0) set_input_file(CNULL);
  gs = (Graphs) make_list(sizeof(struct graphs));
  set_comment(comments);
  new_graphs(gs);
  edit_graphs(gs);
  process_graphs(gs);
  if (show) show_graphs(gs); else draw_graphs(gs, pp, landscape);
  exit(0);
  return 0;
}


